src/hotspot/share/opto/node.cpp

```diff
@@ -903,10 +911,17 @@ void Node::disconnect_inputs(Compile* C) {
 
   // Remove precedence edges if any exist
   // Note: Safepoints may have precedence edges, even during parsing
-  for (uint i = req(); i < len(); ++i) {
-    set_prec(i, nullptr);
+  for (uint i = len() - 1; i < len() && i >= req(); --i) {
```

Comment:
i < len() check is not needed

Reply:
I got now why you need to scan reverse.

Reply:
hi, @vnkozlov ,
Thank you to review it.  

i <len() is there for protection.  Because is unsigned int, len() - 1 or --i might be underflow when len() 0 or req() is 0.

Reply:
I think you could write it like this to hoist the `i < len()` test from the loop while avoiding issues when `len()` or `req()` is 0:

```
uint i = len();
if (i > 0) {
  while(i > req()) {
    rm_prec(--i);
  }
}
```

Reply:
Nice suggestion!

Reply:
hi, @cl4es,  that's `--i` is brilliant! I also tried to hoist comparison out of loop,  but gave up because I feel it makes code less readable. 

we can it less that.  if(i>0) can be saved 
  for (uint i = len(); i > req(); ) {
    rm_prec(--i);  // no-op if _in[i] is nullptr
  }

> "where's the if(len() > 0)" I thought for a second, but you're right: len() >= req(), so if len() is 0, then req() must be 0, too, which means len() > req() is false and we won't accidentally call rm_prec(UINT_MAX)

Yes, that's why I think it's safe to save if (i > 0).  

Actually, I tried to hoist some corner cases out of loop before, but that made code less straightforward.
I give it up because I don't want to win some "performance" at expense of readability and gcc/llvm can do job anyway.

Your code rm_prec(--i) is brilliant. thanks! let's see the result.